src/ucp/core/ucp_am.c

```diff
@@ -0,0 +1,699 @@
+
+    if (ucs_unlikely(am_flags & UCT_CB_PARAM_FLAG_DESC)) {
+        return status;
+    }
+    
+    if (status == UCS_OK) {
+        ucp_recv_desc_release(desc);
+        return UCS_OK;
+    } else if (status == UCS_INPROGRESS) {
+        return UCS_OK;
+    } 
+
+    return status;
+}
+
+static ucs_status_t 
+ucp_am_handler_reply(void *am_arg, void *am_data, size_t am_length,
+                     unsigned am_flags)
+{
+    ucp_am_reply_hdr_t *hdr = (ucp_am_reply_hdr_t *) am_data;
+    ucp_worker_h worker     = (ucp_worker_h) am_arg;
+    uint16_t am_id          = hdr->super.am_id;
+    ucp_ep_h reply_ep;
+
+    reply_ep = ucp_worker_get_ep_by_ptr(worker, hdr->ep_ptr);
+ 
+    return ucp_am_handler_common(worker, hdr + 1, sizeof(*hdr),
+                                 am_length - sizeof(*hdr), reply_ep,
+                                 am_id, UCP_RECV_DESC_FLAG_AM_REPLY, 
+                                 am_flags);
+}
+
+static ucs_status_t 
+ucp_am_handler(void *am_arg, void *am_data, size_t am_length,
+               unsigned am_flags)
+{
+    ucp_worker_h worker   = (ucp_worker_h) am_arg;
+    ucp_am_hdr_t *hdr     = (ucp_am_hdr_t *) am_data;
+    uint16_t am_id        = hdr->am_id;
+
+    return ucp_am_handler_common(worker, hdr + 1, sizeof(*hdr),
+                                 am_length - sizeof(*hdr), NULL,
+                                 am_id, UCP_RECV_DESC_FLAG_AM_HDR, 
+                                 am_flags);    
+}
+
+static ucp_am_unfinished_t *
+ucp_am_find_unfinished(ucp_worker_h worker, ucp_ep_h ep, 
+                       ucp_ep_ext_proto_t *ep_ext, 
+                       ucp_am_long_hdr_t *hdr, size_t am_length)
+{
+    ucp_am_unfinished_t *unfinished;
+    
+    ucs_list_for_each(unfinished, &ep_ext->am.started_ams, unfinished) {
+        if (unfinished->msg_id == hdr->msg_id) {
+            return unfinished;
+        }
+    }
+    
+    return NULL;
+}
+
+static ucs_status_t
+ucp_am_handle_unfinished(ucp_worker_h worker,
+                         ucp_am_unfinished_t *unfinished, 
+                         ucp_am_long_hdr_t *long_hdr, 
+                         size_t am_length, ucp_ep_h reply_ep) {
+    uint16_t am_id;
+    ucs_status_t status;
+
+    memcpy(UCS_PTR_BYTE_OFFSET(unfinished->all_data + 1, long_hdr->offset),
+           long_hdr + 1, am_length - sizeof(*long_hdr));
+    unfinished->left -= am_length - sizeof(*long_hdr);
+    if (unfinished->left == 0) {
+        am_id = long_hdr->am_id;
+        status = worker->am_cbs[am_id].cb(worker->am_cbs[am_id].context,
+                                          unfinished->all_data + 1,
+                                          long_hdr->total_size,
+                                          reply_ep,
+                                          UCP_CB_PARAM_FLAG_DATA);
+
+        if (status != UCS_INPROGRESS) {
+            ucs_free(unfinished->all_data);
+        }
+
+        ucs_list_del(&unfinished->unfinished);
```

Comment:
can 'unfinished->unfinished' be single-linked ucs_queue, instead of double-linked ucs_list?